src/mysql_protocol.erl

```diff
@@ -406,26 +408,37 @@ server_version_to_list(ServerVersion) ->
                            SetFoundRows :: boolean()) ->
     {ok, SockModule :: module(), Socket :: term(),
      SeqNum2 :: non_neg_integer()}.
-maybe_do_ssl_upgrade(SockModule0, Socket0, SeqNum1, _Handshake, undefined,
-                     _Database, _SetFoundRows) ->
+maybe_do_ssl_upgrade(_Host, SockModule0, Socket0, SeqNum1, _Handshake,
+                     undefined, _Database, _SetFoundRows) ->
     {ok, SockModule0, Socket0, SeqNum1};
-maybe_do_ssl_upgrade(gen_tcp, Socket0, SeqNum1, Handshake, SSLOpts,
+maybe_do_ssl_upgrade(Host, gen_tcp, Socket0, SeqNum1, Handshake, SSLOpts,
                      Database, SetFoundRows) ->
     Response = build_handshake_response(Handshake, Database, SetFoundRows),
     {ok, SeqNum2} = send_packet(gen_tcp, Socket0, Response, SeqNum1),
-    case ssl_connect(Socket0, SSLOpts, 5000) of
+    case ssl_connect(Host, Socket0, SSLOpts, 5000) of
         {ok, SSLSocket} ->
             {ok, ssl, SSLSocket, SeqNum2};
         {error, Reason} ->
             exit({failed_to_upgrade_socket, Reason})
     end.
 
-ssl_connect(Port, ConfigSSLOpts, Timeout) ->
-    DefaultSSLOpts = [{versions, [tlsv1]}, {verify, verify_peer}],
+ssl_connect(Host, Port, ConfigSSLOpts, Timeout) ->
+    DefaultSSLOpts = [{versions, [tlsv1]}, {verify, verify_peer},
+                      {server_name_indication, get_default_sni(Host)}],
     MandatorySSLOpts = [{active, false}],
     MergedSSLOpts = merge_ssl_options(DefaultSSLOpts, MandatorySSLOpts, ConfigSSLOpts),
     ssl:connect(Port, MergedSSLOpts, Timeout).
 
+-spec get_default_sni(Host) -> disable | Host
+    when Host :: inet:socket_address() | inet:hostname().
+get_default_sni(Host) when is_list(Host) ->
+    case inet:parse_address(Host) of
+        {ok, _} -> disable;
+        {error, einval} -> Host
+    end;
+get_default_sni(_) ->
+    disable.
```

Comment:
Also here, I'm not sure we should use `disable` (unless we match explicitly on `{local, _}` for unix sockets perhaps, but why use TLS for unix sockets?) A `gen_tcp` host can also be an atom, which I think means non-fully qualified domain name.

What I mean is that adding `disable` potentially lowers the security, which I think we shouldn't do by default if there is any doubt. It can be done by the user.

Reply:
Changed in last commit.

The atom form of the hostname, I had to dig around for that a bit, but finally found out that it can (only?) be the local hostname. I think this doesn't make much sense, and it seems to be practically undocumented, so I would guess that nobody would use that.
Other atoms that can be used (although they stem from the `ip_address` type, not `hostname`) are `loopback` and `any`, which I would treat like an IP adress. We don't have many options here, anyway, except maybe defaulting to `"localhost"` in case of `loopback`...